container/doctabs.go

```diff
@@ -0,0 +1,431 @@
+package container
+
+import (
+	"fyne.io/fyne/v2"
+	"fyne.io/fyne/v2/canvas"
+	"fyne.io/fyne/v2/layout"
+	"fyne.io/fyne/v2/theme"
+	"fyne.io/fyne/v2/widget"
+)
+
+// Declare conformity with Widget interface.
+var _ fyne.Widget = (*DocTabs)(nil)
+
+// DocTabs container is used to display various pieces of content identified by tabs.
+// The tabs contain text and/or an icon and allow the user to switch between the content specified in each TabItem.
+// Each item is represented by a button at the edge of the container.
+//
+// Since: 2.1
+type DocTabs struct {
+	widget.BaseWidget
+
+	Items []*TabItem
+
+	CreateTab      func() *TabItem
+	CloseIntercept func(*TabItem)
+	OnClosed       func(*TabItem)
+	OnSelected     func(*TabItem)
+	OnUnselected   func(*TabItem)
+
+	current  int
+	location TabLocation
+
+	popUpMenu *widget.PopUpMenu
+}
+
+// NewDocTabs creates a new tab container that allows the user to choose between various pieces of content.
+//
+// Since: 2.1
+func NewDocTabs(items ...*TabItem) *DocTabs {
+	tabs := &DocTabs{}
+	tabs.ExtendBaseWidget(tabs)
+	tabs.SetItems(items)
+	return tabs
+}
+
+// Append adds a new TabItem to the end of the tab bar.
+func (t *DocTabs) Append(item *TabItem) {
+	t.SetItems(append(t.Items, item))
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (t *DocTabs) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	r := &docTabsRenderer{
+		baseTabsRenderer: baseTabsRenderer{
+			bar:         &fyne.Container{},
+			divider:     canvas.NewRectangle(theme.ShadowColor()),
+			indicator:   canvas.NewRectangle(theme.PrimaryColor()),
+			buttonCache: make(map[*TabItem]*tabButton),
+		},
+		docTabs:  t,
+		scroller: &Scroll{},
+	}
+	r.action = r.buildAllTabsButton()
+	r.create = r.buildCreateTabsButton()
+	r.box = NewHBox(r.create, r.action)
+	var lastX, lastY float32
+	r.scroller.OnScrolled = func(offset fyne.Position) {
+		// FIXME OnScrolled can be called when the offset hasn't changed (#1868)
+		if offset.X == lastX && offset.Y == lastY {
+			return
+		}
+		lastX = offset.X
+		lastY = offset.Y
+		r.updateIndicator(false)
+	}
+	r.updateAllTabs()
+	r.updateCreateTab()
+	r.updateTabs()
+	r.updateIndicator(false)
+	r.applyTheme(t)
+	return r
+}
+
+// Hide hides the widget.
+//
+// Implements: fyne.Widget
```

Comment:
I think this should be `fyne.CanvasObject`

Reply:
Widget encapsulates CanvasObject, prior art;

- internal/driver/glfw/menu_bar.go#64
- internal/driver/glfw/menu_bar_item.go#73
- internal/widget/overlay_container.go#35
- internal/widget/scroller.go#110,241,499
- internal/widget/shadow.go#63
- widget/entry.go#291
- widget/menu.go#147
- widget/menu_item.go#61
- widget/popup_menu.go#47
- widget/select.go#97
- widget/separator.go#41

Reply:
Goodness it looks like we have a lot to tidy up, thanks.
I think RadioGroup may be the only one that references the root interface.

Reply:
Fixed annotation in AppTabs and DocTabs, other instances will need to be addressed in separate PRs